rfcs/20211207-softmax-v2/README.md

```diff
@@ -0,0 +1,192 @@
+} dnnl_alg_kind_t;
+
+/// A descriptor of a Softmax operation.
+typedef struct {
+    /// The kind of primitive. Used for self-identifying the primitive
+    /// descriptor. Must be #dnnl_softmax.
+    dnnl_primitive_kind_t primitive_kind;
+    /// The kind of propagation. Possible values: #dnnl_forward_training and
+    /// #dnnl_forward_inference.
+    dnnl_prop_kind_t prop_kind;
+    /// Source and destination memory descriptor.
+    dnnl_memory_desc_t src_desc;
+    /// Source and Destination of gradient memory descriptor.
+    dnnl_memory_desc_t diff_src_desc;
+    /// The axis along which to perform the softmax.
+    int softmax_axis;
+    /// Softmax algorithm. Possible values: #dnnl_softmax and
+    /// #dnnl_logsoftmax.
+    dnnl_alg_kind_t alg_kind;
+    /// Destination memory descriptor.
+    dnnl_memory_desc_t dst_desc;
+    /// Destination gradient memory descriptor.
+    dnnl_memory_desc_t diff_dst_desc;
+} dnnl_softmax_v2_desc_t;
+
+/// Primitive descriptor query specification
+typedef enum {
+    ...
+    dnnl_query_softmax_v2_d, ///< softmax version 2 descriptor
+    ...
+} dnnl_query_t;
+```
+
+```c
+/// dnnl.h
+
+/// Initializes a descriptor for softmax v2 forward propagation primitive.
+///
+/// @param softmax_desc Output descriptor for a softmax primitive.
+/// @param prop_kind Propagation kind. Possible values are
+///     #dnnl_forward_training and #dnnl_forward_inference.
+/// @param alg_kind Softmax algorithm kind: either #dnnl_softmax, or
+///     #dnnl_logsoftmax.
+/// @param src_desc Source memory descriptor.
+/// @param dst_desc Destination memory descriptor.
+/// @param softmax_axis Axis over which softmax is computed.
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t DNNL_API dnnl_softmax_v2_forward_desc_init(
+        dnnl_softmax_v2_desc_t *softmax_desc, dnnl_prop_kind_t prop_kind,
+        dnnl_alg_kind_t alg_kind, const dnnl_memory_desc_t *src_desc,
+        const dnnl_memory_desc_t *dst_desc, int softmax_axis);
+
+/// Initializes a descriptor for softmax v2 backward propagation primitive.
+///
+/// @param softmax_desc Output descriptor for a softmax primitive.
+/// @param alg_kind Softmax algorithm kind: either #dnnl_softmax, or
+///     #dnnl_logsoftmax.
+/// @param diff_src_desc Diff source memory descriptors.
+/// @param diff_dst_desc Diff destination memory descriptors.
+/// @param dst_desc Destination memory descriptor.
+/// @param softmax_axis Axis over which softmax is computed.
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t DNNL_API dnnl_softmax_v2_backward_desc_init(
+        dnnl_softmax_v2_desc_t *softmax_desc, dnnl_alg_kind_t alg_kind,
+        const dnnl_memory_desc_t *diff_src_desc,
+        const dnnl_memory_desc_t *diff_dst_desc,
+        const dnnl_memory_desc_t *dst_desc, int softmax_axis);
+```
+
+## Open Questions
+
+* Do we want to mark logsoftmax C API as deprecated?
```

Comment:
I think we want. Do you have any concerns with that?
Also, what about the original `softmax`?
Can we deprecate the `logsoftmax` and say that the `softmax` API will be replaced with `softmax_v2` API in the next major release?

Reply:
The only concern is that users with Werror enabled would have to change their codes regardless. This is what stops me from marking it deprecated in header file. Doc mentioning it is absolutely go.

Original softmax should be deprecated as well, with the same message we use for logsoftmax.

Reply:
Yeah, we don't seem to use the practice of deprecating API in headers. I thought it's just about deprecation in general.

Deprecation notice in the documentation and release notes sounds good to me.

Reply:
Put deprecation actions under option text. Thanks for a discussion.